pkg/manager/internal.go

```diff
@@ -190,6 +180,9 @@ type controllerManager struct {
 	// after the gracefulShutdownTimeout ended. It must not be accessed before internalStop
 	// is closed because it will be nil.
 	shutdownCtx context.Context
+
+	internalCtx    context.Context
+	internalCancel context.CancelFunc
```

Comment:
Going to add some documentation on what these are used for internally once we agree this is what we want / have some feedback

Reply:
`Add` is where I hit a blocker, if we pass a context in there, we might not be able to cancel all running routines depending on which context is passed in. 

I had to go with an internal context and force it on every `Add` call to make sure that we cancel that context before exiting.

The other issue I found with having a tree of context structs is that when you do `WithCancel`, the cancel is propagated and we don't really want that for this use case.

I'm happy to iterate on it in multiple small steps, this was the one that was the least confusing 😄

Reply:
Yeah I think this won't be possible without storing it, unless we want to create a new context on every `Add` and then store a slice of cancel funcs that are all called on stop, which seems a lot worse